File: backend/src/utils/secretCipher.ts
import crypto from "crypto";
import { requireEnv } from "./env";

/** Derives a 32-byte AES key from JWT_SECRET (already required for the app to
 * run at all) rather than introducing a second required secret just for
 * encrypting admin-configured values like the SMTP password at rest. */
function deriveKey(): Buffer {
  return crypto.createHash("sha256").update(requireEnv("JWT_SECRET")).digest();
}

/** Encrypts a short secret (e.g. an SMTP password) for storage in a plain
 * key/value settings table. Output is "<iv-b64>.<tag-b64>.<ciphertext-b64>" —
 * comfortably under AdminSetting.value's 255-char column for any realistic
 * password. */
export function encryptSecret(plain: string): string {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", deriveKey(), iv);
  const encrypted = Buffer.concat([cipher.update(plain, "utf8"), cipher.final()]);
  const authTag = cipher.getAuthTag();
  return [iv, authTag, encrypted].map((b) => b.toString("base64")).join(".");
}

export function decryptSecret(payload: string): string {
  const [ivB64, tagB64, dataB64] = payload.split(".");
  const decipher = crypto.createDecipheriv("aes-256-gcm", deriveKey(), Buffer.from(ivB64, "base64"));
  decipher.setAuthTag(Buffer.from(tagB64, "base64"));
  return Buffer.concat([decipher.update(Buffer.from(dataB64, "base64")), decipher.final()]).toString("utf8");
}

File: backend-py/app/security/secret_cipher.py
"""Port of `backend/src/utils/secretCipher.ts` — AES-256-GCM encrypt/decrypt
for short secrets (e.g. an SMTP password) stored in the `AdminSettings`
key/value table (TODO(phase N): the settings service that actually reads/
writes these). The key is derived from `JWT_SECRET` (SHA-256), the same
choice the Node side made to avoid a second required secret.

Output format: "<iv-b64>.<tag-b64>.<ciphertext-b64>" — identical to the TS
implementation, so a value encrypted by one backend decrypts correctly on
the other (both derive the same key from the same `JWT_SECRET`).
"""

from __future__ import annotations

import base64
import hashlib
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from app.config import settings

_IV_LENGTH_BYTES = 12


def _derive_key() -> bytes:
    return hashlib.sha256(settings.require_jwt_secret().encode("utf-8")).digest()


def encrypt_secret(plain: str) -> str:
    key = _derive_key()
    iv = os.urandom(_IV_LENGTH_BYTES)
    aesgcm = AESGCM(key)
    # `cryptography`'s AESGCM appends the 16-byte auth tag to the ciphertext;
    # split it back out so the on-the-wire format matches Node's
    # iv/tag/ciphertext triple exactly.
    ciphertext_and_tag = aesgcm.encrypt(iv, plain.encode("utf-8"), None)
    ciphertext, tag = ciphertext_and_tag[:-16], ciphertext_and_tag[-16:]
    return ".".join(
        base64.b64encode(part).decode("ascii") for part in (iv, tag, ciphertext)
    )


def decrypt_secret(payload: str) -> str:
    iv_b64, tag_b64, data_b64 = payload.split(".")
    key = _derive_key()
    iv = base64.b64decode(iv_b64)
    tag = base64.b64decode(tag_b64)
    ciphertext = base64.b64decode(data_b64)
    aesgcm = AESGCM(key)
    plain = aesgcm.decrypt(iv, ciphertext + tag, None)
    return plain.decode("utf-8")
